feat: non-blocking rate limiting on refresh.

hessjcg opened the pull request:
This further improve the efficient use of futures in the refresh by using the ScheduledExecutor's delay function to
rate limit instead of blocking the ScheduleExecutor's execution thread.

enocom commented:
Let's get these tests green before we review.

hessjcg commented on core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:
```diff
@@ -220,25 +219,19 @@ private ListenableFuture<InstanceData> startRefreshAttempt() {
-                  String.format(
-                      "[%s] Refresh Operation: Acquiring rate limiter permit.", instanceName));
-              //noinspection UnstableApiUsage
-              forcedRenewRateLimiter.acquire();
```
We're removing this because it blocks an executor thread.

hessjcg commented on core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:
```diff
@@ -220,25 +219,19 @@ private ListenableFuture<InstanceData> startRefreshAttempt() {
-            executor);
+    logger.fine(
+        String.format("[%s] Refresh Operation: Acquiring rate limiter permit.", instanceName));
+    ListenableFuture<Long> delay = rateLimitCalculator.acquireAsync(executor);
```
... and replacing it with this because it uses executor.schedule() to wait without blocking an executor thread.

enocom commented on core/src/main/java/com/google/cloud/sql/core/RateLimitCalculator.java:
```diff
@@ -0,0 +1,67 @@
+   * If this returns <= 0, the operation must call "acquire" again until it returns -1 to avoid race
+   * conditions.
+   */
+  synchronized long acquire() {
```
Shouldn't this be a private method?

enocom replied:
Can we test only through the public interface? Would a deterministic scheduler help do this? We should try to avoid testing private interfaces.

hessjcg replied:
Unfortunately, this is one of those rare situations where we need to unit test a private method because: 

- there are detailed states that we must test precisely to ensure that the rate limiter works.
- the rate limiter's public API does not grant the precision we need to create those states reliably. 
 
I can't write a precise test case using the public API `acquireAsync()` to validate that the delay calculation logic is working properly. Timing in a real thread pool is too unreliable. The jmock DeterministicScheduler doesn't work correctly with scheduled tasks in a ListeningScheduledExecutorService.

enocom commented on core/src/test/java/com/google/cloud/sql/core/RateLimitCalculatorTest.java:
```diff
@@ -0,0 +1,134 @@
+  @Test
+  public void firstCallShouldReturnNoDelay() {
+    RateLimitCalculator c = new RateLimitCalculator(100);
+    assertThat(c.acquire()).isEqualTo(-1);
```
Are we testing a private method here?

hessjcg commented on core/src/main/java/com/google/cloud/sql/core/RateLimitCalculator.java:
```diff
@@ -0,0 +1,67 @@
+ * A simple, constant-time rate limit calculator. Ensures that there is always at least
+ * delayBetweenAttempts milliseconds between attempts.
+ */
+class RateLimitCalculator {
```
Ok, so I think we call this RateLimiter, since the primary public interface is `acquireAsync()`.

hessjcg commented on core/src/main/java/com/google/cloud/sql/core/RateLimitCalculator.java:
```diff
@@ -0,0 +1,67 @@
+          () -> this.acquireAsync(executor), limit, TimeUnit.MILLISECONDS, executor);
+    }
+
+    System.out.println("Acquired! " + limit);
```
Removed. This was left over from debugging.

hessjcg commented on core/src/main/java/com/google/cloud/sql/core/CoreSocketFactory.java:
```diff
@@ -410,6 +409,6 @@ private CloudSqlInstance apiFetcher(ConnectionConfig config) {
         executor,
         localKeyPair,
-        RateLimiter.create(1.0 / 30.0)); // 1 refresh attempt every 30 seconds
+        30 * 1000); // 1 refresh attempt every 30 seconds
```
Fixed.

enocom reviewed (commented):
Why did we add a RateLimiterTracker?

hessjcg commented on core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:
```diff
@@ -89,13 +86,15 @@ class CloudSqlInstance {
-    this.forcedRenewRateLimiter = forcedRenewRateLimiter;
+
+    // convert requests/second into milliseconds between requests.
+    this.rateLimitCalculator = new AsyncRateLimiter(minRefreshDelayMs);
```
Fixed.

hessjcg commented on core/src/main/java/com/google/cloud/sql/core/AsyncRateLimiter.java:
```diff
@@ -0,0 +1,65 @@
+   *
+   * @param executor the executor to use to schedule future checks for available rate limits.
+   */
+  public ListenableFuture<Long> acquireAsync(ScheduledExecutorService executor) {
```
I've switched this to `ListenableFuture<Void>`. Our linter throws a warning when we use `ListenableFuture` because it doesn't allow us to use a parameterized class without type parameters.

hessjcg replied:
Actually, I had to switch back to making an internal class to track when the rate limit was acquired. Now it's called `RateLimitAcquisition` and it is package-scope and labeled with `@VisibileForTesting`. 

Unfortunately, there is way too much variability in the thread executors across all the platforms to test this with the public api. Measuring the time that acquireAsync() resolved using `Future.addListener()` in the test case introduced much delay on Windows and sometimes Mac VMs. This make the tests flakey.

hessjcg commented:
> Why did we add a RateLimiterTracker?
I removed `RateLimitTracker` and rewrote the test case.

I attempted to use the DeterministicScheduler, but it seems that it does not play nicely with scheduled tasks. 

```java
public class AsyncRateLimiterTest {

  @Test
  public void determinisicScheduleTest() {
    final long delay = 200;
    AsyncRateLimiter c = new AsyncRateLimiter(delay);

    DeterministicScheduler ex = new DeterministicScheduler();

    List<ListenableFuture<?>> futures = new ArrayList<>();
    for (int i = 0; i < 3; i++) {
      futures.add(c.acquireAsync(ex));
    }

    // First attempt happens without any delay, because nothing is run yet.
    assertThat(futures.stream().mapToInt(f -> f.isDone() ? 1 : 0).sum()).isEqualTo(1);

    // Tick forward less than the delay.
    ex.tick(50, TimeUnit.MILLISECONDS);

    // When all futures are evaluated again immediately, still only 1 request has finished.
    // because not enough time has elapsed.
    assertThat(futures.stream().mapToInt(f -> f.isDone() ? 1 : 0).sum()).isEqualTo(1);

    // Tick forward more than the delay. Now 2 attempts should have finished.
    ex.tick(delay, TimeUnit.MILLISECONDS);
    ex.runUntilIdle();

    // Now wait until the delay has expired and evaluate again. Now, 2 attempts should have
    // finished.
    assertThat(futures.stream().mapToInt(f -> f.isDone() ? 1 : 0).sum()).isEqualTo(2); // THIS FAILS, wants: 2, got: 1
  }
}
```

enocom commented:
Let's pair on this a bit to see if we can get a more testable interface without exposing private internals.
